### PR TITLE
feat: remove the state in the member cluster CRD

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

All the existing UT/IT/E2E should work

### Special notes for your reviewer

I am not sure the mc watcher for scheduler need to enqueue all the CRPs when a cluster left the fleet.

### CODE REVIEW

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 1]
```diff
@@ -194,14 +193,8 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 			clusterKObj := klog.KObj(newCluster)
-			if newCluster.Spec.State == clusterv1beta1.ClusterStateJoin && !reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
-				klog.V(2).Info("An member cluster label change has been detected", "memberCluster", clusterKObj)
-				return true
-			}
-
-			// The cluster has left.
-			if oldCluster.Spec.State == clusterv1beta1.ClusterStateJoin && newCluster.Spec.State == clusterv1beta1.ClusterStateLeave {
-				klog.V(2).Info("A member cluster has left the fleet", "memberCluster", clusterKObj)
+			if !reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
```
Comment (zhiying-lin) [inline pkg/scheduler/watchers/membercluster/controller.go:196]: check if the deletiontimestamp is not added in the newCluster?

File: pkg/controllers/membercluster/v1beta1/membercluster_controller.go [hunk 2]
```diff
@@ -128,6 +104,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// calculate the current status of the member cluster from imc status
+	r.syncInternalMemberClusterStatus(currentImc, mc)
+	cond := meta.FindStatusCondition(mc.Status.Conditions, string(clusterv1beta1.AgentJoined))
+	// cluster never joined or already left
+	if cond != nil && cond.Status == metav1.ConditionFalse && cond.ObservedGeneration == mc.GetGeneration() {
+		klog.V(2).InfoS("No need to wait for agent leave, start garbage collecting", "memberCluster", mcObjRef)
+		return r.garbageCollectWork(ctx, mc)
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/controllers/membercluster/v1beta1/membercluster_controller.go:125]: No, we can't. The GC will remove all the works/imc/namespace

File: pkg/controllers/membercluster/v1beta1/membercluster_controller.go [hunk 3]
```diff
@@ -128,6 +104,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	}
+
+	// mark the imc as left again to make sure the agent is leaving the fleet
+	if err := r.leave(ctx, mc, currentImc); err != nil {
+		klog.ErrorS(err, "failed to leave", "memberCluster", mcObjRef)
+		return ctrl.Result{}, err
+	}
+	// update the mc status while we wait for all the agents to leave
+	err = r.updateMemberClusterStatus(ctx, mc)
+	return ctrl.Result{}, controller.NewUpdateIgnoreConflictError(err)
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/controllers/membercluster/v1beta1/membercluster_controller.go:135]: the finalizer is removed after the GC work is done

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 4]
```diff
@@ -169,10 +169,9 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 			return false
 		},
 		DeleteFunc: func(e event.DeleteEvent) bool {
-			// Ignore deletion events; the removal of a cluster is first signaled by the join
-			// state change (join -> leave), which is already handled separately.
-			klog.V(2).Info("Ignoring delete events for member cluster objects", "eventObject", klog.KObj(e.Object))
-			return false
+			// the cluster is leaving the fleet
+			klog.V(2).Info("delete events for member cluster objects", "eventObject", klog.KObj(e.Object))
+			return true
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/scheduler/watchers/membercluster/controller.go:174]: the watcher needs to handle cluster "leave"

File: pkg/scheduler/watchers/membercluster/controller.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/scheduler/watchers/membercluster/controller.go:196]: one can't update a "deleting" object so it can't have deletionTimestamp.

Comment (zhiying-lin) [reply pkg/scheduler/watchers/membercluster/controller.go:196]: Deleting a resource means add a deletiontimestamp.
 https://github.com/Azure/fleet/blob/main/pkg/scheduler/watchers/clusterresourceplacement/controller.go#L75-L89 which has been validated by integration tests.

File: pkg/scheduler/watchers/membercluster/controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/scheduler/watchers/membercluster/controller.go:174]: learn from Chen's PR, the delete event means, when the object is actually removed from storage, https://github.com/Azure/fleet/blob/main/pkg/scheduler/watchers/clusterresourceplacement/controller.go#L71

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 5]
```diff
@@ -169,10 +169,9 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 		},
 		UpdateFunc: func(e event.UpdateEvent) bool {
```
Comment (michaelawyu) [inline pkg/scheduler/watchers/membercluster/controller.go:176]: Hi Ryan! For this one we also need to enqueue when the deletion timestamp is added (updated) to the new object; the system considers the deletion op as an update event first (ObjectMeta updated) then a delete event (object removed).

**Comment by michaelawyu** [comment]
Other than the one comment things LGTM ;)

File: apis/v1alpha1/types.go [hunk 6]
```diff
@@ -1,18 +1,15 @@
...
 )
 
-// A Conditioned may have conditions set or retrieved. Conditions typically
-// indicate the status of both a resource and its reconciliation process.
-// +kubebuilder:object:generate=false
-type Conditioned interface {
-	SetConditions(...metav1.Condition)
-	GetCondition(string) *metav1.Condition
-}
-
```
Comment (zhiying-lin) [inline apis/v1alpha1/types.go:15]: also need to remove the interfaces in apis/placement/v1beta1/types.go

Comment (ryanzhang-oss) [reply apis/v1alpha1/types.go:15]: I can't seem to find that file.

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 7]
```diff
@@ -192,16 +192,15 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+			if oldCluster.GetDeletionTimestamp().IsZero() && !newCluster.GetDeletionTimestamp().IsZero() {
+				klog.V(2).InfoS("A member cluster is leaving the fleet", "memberCluster", clusterKObj)
 				return true
 			}
-
-			// The cluster has left.
-			if oldCluster.Spec.State == clusterv1beta1.ClusterStateJoin && newCluster.Spec.State == clusterv1beta1.ClusterStateLeave {
-				klog.V(2).InfoS("A member cluster has left the fleet", "memberCluster", clusterKObj)
+			// Note that the controller runs only when label changes happen on joined clusters.
+			if !reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
```
Comment (zhiying-lin) [inline pkg/scheduler/watchers/membercluster/controller.go:202]:
```suggestion
			if newCluster.GetDeletionTimestamp().IsZero() && !reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
```

i tested that the labels still can be updated in a cr with deletiontimesamp.

Comment (zhiying-lin) [reply pkg/scheduler/watchers/membercluster/controller.go:202]: confirmed with Chen and scheduler will check the deletionTimestamp again.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**